app/models/topic_list.rb

```diff
@@ -113,6 +113,8 @@ def load_topics
       user_ids << ft.user_id << ft.last_post_user_id << ft.featured_user_ids << ft.allowed_user_ids
     end
 
+    DiscourseEvent.trigger(:before_topic_list_user_lookup, @topics, user_ids)
```

Comment:
I've always found `DiscourseEvent` to be a bit of a hack because it doesn't allow us to clearly define what the extension is supposed to do. In this case, `user_ids` is being mutated by the plugin which may be surprising when debugging an problem. Instead of a trigger, can we do something like what we do for `TopicView` where we define a clear API for plugins to follow. 

https://github.com/discourse/discourse/blob/f03f0866e758cbd0b98a087b9e10cfc3a70ef9d8/lib/topic_view.rb#L73-L76
https://github.com/discourse/discourse/blob/f03f0866e758cbd0b98a087b9e10cfc3a70ef9d8/lib/plugin/instance.rb#L205-L211

Reply:
That makes sense. I updated the PR with the new plugin API.